<?xml version="1.0" ?>
<robot name="sensor_realsense_r200" xmlns:xacro="http://ros.org/wiki/xacro">

     <xacro:macro name="sensor_realsense_r200" params="name:='' prefix:='' parent *origin visualize:=false">
          <!-- Camera link -->
          <link name="${prefix}r200_camera">
               <visual>
                    <origin rpy="0 0 0" xyz="0 0 0"/>
                    <geometry>
                         <mesh filename="package://gazebo_sensor_collection/meshes/cameras/intel/realsense_r200.dae"/>
                    </geometry>
               </visual>
               <collision>
                    <geometry>
                         <box size="0.0078 0.130 0.0192"/>
                    </geometry>
               </collision>
               <inertial>
                    <mass value="0.0615752"/>
                    <origin rpy="0 0 0" xyz="0 0 0"/>
                    <inertia ixx="9.108e-06"
                         ixy="0"
                         ixz="0"
                         iyy="2.51e-07"
                         iyz="0"
                         izz="8.931e-06"
                         />
               </inertial>
          </link>

          <!-- Connect camera to parent frame-->
          <joint name="${prefix}realsense_joint" type="fixed">
               <parent link="${parent}"/>
               <child link="${prefix}r200_camera"/>
               <xacro:insert_block name="origin"/>
          </joint>

          <!-- Virtual links representing the cameras positons and orientations-->
          <link name="${prefix}color" />
          <link name="${prefix}depth" />
          <link name="${prefix}ired1" />
          <link name="${prefix}ired2" />

          <!-- Joints positioning the virtual links with respect to the camera main link-->
          <joint name="${prefix}color_joint" type="fixed">
               <parent link="${prefix}r200_camera" />
               <child link="${prefix}color" />
               <origin xyz="0 -0.046 0.004" rpy="${pi/2} ${pi} ${pi/2}"/>
          </joint>

          <joint name="${prefix}depth_joint" type="fixed">
               <parent link="${prefix}r200_camera" />
               <child link="${prefix}depth" />
               <origin xyz="0 -0.03 0.004" rpy="0 0 0"/>
          </joint>

          <joint name="${prefix}ired1_joint" type="fixed">
               <parent link="${prefix}r200_camera" />
               <child link="${prefix}ired1" />
               <origin xyz="0 -0.06 0.004" rpy="0 0 0"/>
          </joint>

          <joint name="${prefix}ired2_joint" type="fixed">
               <parent link="${prefix}r200_camera" />
               <child link="${prefix}ired2" />
               <origin xyz="0 0.01 0.004" rpy="0 0 0"/>
          </joint>

          <!-- ====================================================
                                   Gazebo
          ========================================================= -->
          <gazebo>
               <plugin name="realsense_plugin" filename="libgazebo_ros_realsense.so">
                    <cameraName>r200</cameraName>
                    <baseline>70</baseline>            <!-- in mm -->
                    <depthNearClip>0.3</depthNearClip> <!-- in m -->
                    <depthFarClip>10.0</depthFarClip>  <!-- in m -->
                    <depthScale>0.001</depthScale>
               </plugin>
          </gazebo>

          <!-- Load parameters to model's main link-->
          <gazebo reference="${prefix}r200_camera">
               <!-- ================== Color Camera =================== -->
               <sensor name="color" type="camera">
                    <pose frame="">0 -0.046 0.004 0 0 0</pose>
                    <camera name="__default__">
                         <horizontal_fov>1.22173</horizontal_fov>
                         <image>
                              <width>640</width>
                              <height>480</height>
                              <format>RGB_INT8</format>
                         </image>
                         <clip>
                              <near>0.1</near>
                              <far>100</far>
                         </clip>
                         <noise>
                              <type>gaussian</type>
                              <mean>0.0</mean>
                              <stddev>0.007</stddev>
                         </noise>
                    </camera>
                    <always_on>1</always_on>
                    <update_rate>60</update_rate>
                    <visualize>${visualize}</visualize>
               </sensor>
               <!-- ============== Infared Camera 1 =================== -->
               <sensor name="ired1" type="camera">
                    <pose frame="">0 -0.06 0.004 0 0 0</pose>
                    <camera name="__default__">
                         <horizontal_fov>1.02974</horizontal_fov>
                         <image>
                              <width>640</width>
                              <height>480</height>
                              <format>L_INT8</format>
                         </image>
                         <clip>
                              <near>0.1</near>
                              <far>100</far>
                         </clip>
                         <noise>
                              <type>gaussian</type>
                              <mean>0.0</mean>
                              <stddev>0.05</stddev>
                         </noise>
                    </camera>
                    <always_on>1</always_on>
                    <update_rate>60</update_rate>
                    <visualize>0</visualize>
               </sensor>
               <!-- ============== Infared Camera 2 =================== -->
               <sensor name="ired2" type="camera">
                    <pose frame="">0 0.01 0.004 0 0 0</pose>
                    <camera name="__default__">
                         <horizontal_fov>1.02974</horizontal_fov>
                         <image>
                              <width>640</width>
                              <height>480</height>
                              <format>L_INT8</format>
                         </image>
                         <clip>
                              <near>0.1</near>
                              <far>100</far>
                         </clip>
                         <noise>
                              <type>gaussian</type>
                              <mean>0.0</mean>
                              <stddev>0.05</stddev>
                         </noise>
                    </camera>
                    <always_on>1</always_on>
                    <update_rate>60</update_rate>
                    <visualize>0</visualize>
               </sensor>
               <!-- ================= Depth Camera =================== -->
               <sensor name="depth" type="depth">
                    <pose frame="">0 -0.03 0.004 0 0 0</pose>
                    <camera name="__default__">
                         <horizontal_fov>1.02974</horizontal_fov>
                         <image>
                              <width>640</width>
                              <height>480</height>
                         </image>
                         <clip>
                              <near>0.1</near>
                              <far>100</far>
                         </clip>
                         <noise>
                              <type>gaussian</type>
                              <mean>1.0</mean>
                              <stddev>0.5</stddev>
                         </noise>
                    </camera>
                    <always_on>1</always_on>
                    <update_rate>60</update_rate>
                    <visualize>0</visualize>
               </sensor>
          </gazebo>
     </xacro:macro>

</robot>
